### PR TITLE
Fix Issue 34748 - read in datetime as MultiIndex for column headers

### LINKED ISSUE #34748: BUG: read_excel issue - 2 (when "column name" is a date/time)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
pd.read_excel('test.xlsx', header=[0, 1], index_col=0, engine='openpyxl')
```
(`engine` could be `xlrd` or `openpyxl`, doesn't matter.)

[test.xlsx](https://github.com/pandas-dev/pandas/files/4774723/test.xlsx)

#### Problem description

Exception:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    pd.read_excel('test.xlsx', header=[0, 1], index_col=0, engine='xlrd')
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 334, in read_excel
    **kwds,
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 888, in parse
    **kwds,
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 512, in parse
    **kwds,
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2201, in TextParser
    return TextFileReader(*args, **kwds)
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 880, in __init__
    self._make_engine(self.engine)
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 1126, in _make_engine
    self._engine = klass(self.f, **self.options)
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2298, in __init__
    self.columns, self.index_names, self.col_names
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 1508, in _extract_multi_indexer_columns
    for r in header
  File "/home/sasha/miniconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 1508, in <listcomp>
    for r in header
TypeError: object of type 'datetime.datetime' has no len()
```

#### Expected Output

No exceptions.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.42-calculate
machine          : x86_64
processor        : Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
byteorder        : little
LC_ALL           : None
LANG             : ru_RU.utf8
LOCALE           : ru_RU.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.4.0.post20200518
Cython           : 0.29.17
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.1.3
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.2.8
numba            : None
</details>

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] closes #34748
- [x] tests added / passed
  - `pandas/tests/io/excel/test_readers.py`, method `test_read_datetime_multiindex`
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by WillAyd — COMMENTED** [review]
looks good - can you test for the other file extensions?

**Comment by Dr-Irv** [comment]
> looks good - can you test for the other file extensions?

Just pushed a change. Couldn't figure out how to get an ODS file that would successfully read, so I put in xfail for that

**Comment by WillAyd** [comment]
@Dr-Irv added on ODS file to this if you want to pull locally

**Comment by Dr-Irv** [comment]
@WillAyd Thanks.  Latest push merges with master and removes the xfail on ods.  Passes on my Windows box....

**Comment by Dr-Irv** [comment]
@WillAyd I'm currently failing because the XLSB file I created via Excel can't be read by pandas using `xlrd` or `openpyxl`.  Is there a way to create xlsb files for testing that pandas can read?

**Comment by WillAyd** [comment]
pyxlsb is the only engine that can read that format. IIRC there’s also a limitation in that tool where it can’t distinguish datetime types which may be problematic for this (can skip until that gets patched if needed)

**Comment by Dr-Irv** [comment]
> pyxlsb is the only engine that can read that format. IIRC there’s also a limitation in that tool where it can’t distinguish datetime types which may be problematic for this (can skip until that gets patched if needed)

Thanks @WillAyd .  pyxlsb not supporting datetimes was the issue.

**Comment by WillAyd** [comment]
ref #31215 and https://github.com/willtrnr/pyxlsb/issues/10

File: pandas/tests/io/excel/test_readers.py
Comment (WillAyd) [inline pandas/tests/io/excel/test_readers.py:1158]: Are the to_pydatetime calls required?

**Review by WillAyd — COMMENTED** [review]
Cool thanks! Implementation lgtm minor question on test

@jreback

File: pandas/tests/io/excel/test_readers.py
Comment (jreback) [inline pandas/tests/io/excel/test_readers.py:1162]: what does this comment mean?

**Review by jreback — CHANGES_REQUESTED** [review]
2 questions, patch and tests look good

File: pandas/tests/io/excel/test_readers.py
Comment (Dr-Irv) [reply pandas/tests/io/excel/test_readers.py:1158]: Because when the Excel reader creates the names of the index, the types are of `dt.datetime` not the pandas `datetime`

File: pandas/tests/io/excel/test_readers.py
Comment (Dr-Irv) [reply pandas/tests/io/excel/test_readers.py:1162]: So there is a bug in master (but not in 1.0.5).  See below.  Should I create a separate issue for this?
```

In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.1.0.dev0+1952.gc744e35d0.dirty'

In [3]: expected_column_index = pd.MultiIndex.from_tuples(
   ...:     [(pd.to_datetime("02/29/2020"), pd.to_datetime("03/01/2020"))],
   ...:     names=[
   ...:         pd.to_datetime("02/29/2020").to_pydatetime(),
   ...:         pd.to_datetime("03/01/2020").to_pydatetime(),
   ...:     ],
   ...: )

In [4]: pd.DataFrame([], columns=expected_column_index)
---------------------------------------------------------------------------
InvalidIndexError                         Traceback (most recent call last)
<ipython-input-4-e63e2d18b57f> in <module>
----> 1 pd.DataFrame([], columns=expected_column_index)

c:\Code\pandas_dev\pandas\pandas\core\frame.py in __init__(self, data, index, columns, dtype, copy)
    515                     mgr = init_ndarray(data, index, columns, dtype=dtype, copy=copy)
    516             else:
--> 517                 mgr = init_dict({}, index, columns, dtype=dtype)
    518         else:
    519             try:

c:\Code\pandas_dev\pandas\pandas\core\internals\construction.py in init_dict(data, index, columns, dtype)
    267                 nan_dtype = dtype
    268             val = construct_1d_arraylike_from_scalar(np.nan, len(index), nan_dtype)
--> 269             arrays.loc[missing] = [val] * missing.sum()
    270
    271     else:

c:\Code\pandas_dev\pandas\pandas\core\indexing.py in __setitem__(self, key, value)
    662         else:
    663             key = com.apply_if_callable(key, self.obj)
--> 664         indexer = self._get_setitem_indexer(key)
    665         self._has_valid_setitem_indexer(key)
    666

c:\Code\pandas_dev\pandas\pandas\core\indexing.py in _get_setitem_indexer(self, key)
    613
    614         try:
--> 615             return self._convert_to_indexer(key, axis=0, is_setter=True)
    616         except TypeError as e:
    617

c:\Code\pandas_dev\pandas\pandas\core\indexing.py in _convert_to_indexer(self, key, axis, is_setter)
   1158             # if we are a label return me
   1159             try:
-> 1160                 return labels.get_loc(key)
   1161             except LookupError:
   1162                 if isinstance(key, tuple) and isinstance(labels, ABCMultiIndex):

c:\Code\pandas_dev\pandas\pandas\core\indexes\multi.py in get_loc(self, key, method)
   2694         if not isinstance(key, (tuple, list)):
   2695             # not including list here breaks some indexing, xref #30892
-> 2696             loc = self._get_level_indexer(key, level=0)
   2697             return _maybe_to_slice(loc)
   2698

c:\Code\pandas_dev\pandas\pandas\core\indexes\multi.py in _get_level_indexer(self, key, level, indexer)
   2959         else:
   2960
-> 2961             code = self._get_loc_single_level_index(level_index, key)
   2962
   2963             if level > 0 or self.lexsort_depth == 0:

c:\Code\pandas_dev\pandas\pandas\core\indexes\multi.py in _get_loc_single_level_index(self, level_index, key)
   2628             return -1
   2629         else:
-> 2630             return level_index.get_loc(key)
   2631
   2632     def get_loc(self, key, method=None):

c:\Code\pandas_dev\pandas\pandas\core\indexes\datetimes.py in get_loc(self, key, method, tolerance)
    569         """
    570         if not is_scalar(key):
--> 571             raise InvalidIndexError(key)
    572
    573         orig_key = key

InvalidIndexError: 2020-02-29 00:00:00  2020-03-01 00:00:00
2020-02-29           2020-03-01             True
dtype: bool
```

File: pandas/tests/io/excel/test_readers.py
Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1158]: hmm, where is this done? this is unfortunately as these should actually be Timestamp

File: pandas/tests/io/excel/test_readers.py
Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1162]: see above, i think that should work, but these should be coerced to Timestamp, we *don't* support datetime in Index at all (even as object), they are *always* coerced.

Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1162]: would be ok with handling this as a pre-cursor PR bug fix i think.

Comment (Dr-Irv) [reply pandas/tests/io/excel/test_readers.py:1162]: > see above, i think that should work, but these should be coerced to Timestamp, we _don't_ support datetime in Index at all (even as object), they are _always_ coerced.

In my example, the index values are `TimeStamp`, but the Excel reader uses `datetime` for the *names*

Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1162]: right, how does the Excel reader do that?

File: pandas/tests/io/excel/test_readers.py
Comment (Dr-Irv) [reply pandas/tests/io/excel/test_readers.py:1158]: > hmm, where is this done? this is unfortunately as these should actually be Timestamp

So is this a separate issue - that we don't want the *names* to be dt.datetime ?  If so, I will create an issue for that.

File: pandas/tests/io/excel/test_readers.py
Comment (Dr-Irv) [reply pandas/tests/io/excel/test_readers.py:1162]: > would be ok with handling this as a pre-cursor PR bug fix i think.

So do you mean we hold up on this PR, create an issue for the bug above, and get that fixed first?

**Comment by Dr-Irv** [comment]
@jreback Summarizing the discussion above.  There seem to be 2 separate things raised above:

1. In current master, if you use `TimeStamp` as a `MultiIndex` for the columns, there is a failure:
```python
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.1.0.dev0+1952.gc744e35d0.dirty'

In [3]: expected_column_index = pd.MultiIndex.from_tuples(
   ...:     [(pd.to_datetime("02/29/2020"), pd.to_datetime("03/01/2020"))],
   ...:     names=[
   ...:         'a', 'b'])

In [4]: pd.DataFrame([], columns=expected_column_index)
---------------------------------------------------------------------------
InvalidIndexError
```
2. The Excel reader (and in fact all constructors) allow the name of the column index to be a `datetime.datetime`:
```python
In [1]: import pandas as pd

In [2]: import datetime as dt

In [3]: df=pd.DataFrame([], columns=pd.Index(["a"], name=dt.datetime(2020,6,25)
   ...: ))

In [4]: df.columns
Out[4]: Index(['a'], dtype='object', name=2020-06-25 00:00:00)

In [5]: type(df.columns.name)
Out[5]: datetime.datetime
```

It seems to me you are suggesting that I create an issue for (1) and we fix that before merging in this PR.

For (2), I suggest that we don't worry about that, or we could create an issue that says that pandas needs to coerce the names of indexes that are `datetime.datetime` into `Timestamp`

Let me know if you'd like to proceed in a different manner.

File: pandas/tests/io/excel/test_readers.py
Comment (jreback) [reply pandas/tests/io/excel/test_readers.py:1162]: so if the other issue can be address first would be best. so i would create an issue. ideally we merge the patch first. if it turns into a quagmire, then can re-visit.

**Comment by Dr-Irv** [comment]
@jreback pinging you here, as now that I fixed the other bug, this PR should be ready for merge

**Comment by jreback** [comment]
> @jreback pinging you here, as now that I fixed the other bug, this PR should be ready for merge

don't we need to change to use Timestamps?

**Comment by Dr-Irv** [comment]
> > @jreback pinging you here, as now that I fixed the other bug, this PR should be ready for merge
> 
> don't we need to change to use Timestamps?

See discussion above (https://github.com/pandas-dev/pandas/pull/34954#issuecomment-649614583).   That's a separate issue (# 2 in that comment, which I could create if you want), which is whether or not we want to allow the _names_ of a `MultiIndex` to be `datetime` objects, which would also entail a change to the Excel reader (and maybe elsewhere), so that a name can't be a `datetime` but is coerced to a `Timestamp`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Dr-Irv ok i get it

**Pull request merged**
